test/test_distributions.py

```diff
@@ -958,6 +872,217 @@ def test_invalid_parameter_broadcasting(self):
+                'actual {}'.format(actual_grad),
+                'rel error {}'.format(rel_error),
+                'max error {}'.format(rel_error.max()),
+                'at x = {}'.format(x[rel_error.argmax()]),
+            ]))
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_beta_wrt_beta(self):
+        num_samples = 20
+        grid = [1e-2, 1e-1, 1e0, 1e1, 1e2]
+        for alpha, beta in product(grid, grid):
+            betas = Variable(torch.FloatTensor([beta] * num_samples), requires_grad=True)
+            alphas = Variable(torch.FloatTensor([alpha] * num_samples).type_as(betas))
+            x = Beta(alphas, betas).rsample()
+            x.sum().backward()
+            x, ind = x.data.sort()
+            x = x.numpy()
+            actual_grad = betas.grad.data[ind].numpy()
+            # Compare with expected gradient dx/dbeta along constant cdf(x,alpha,beta).
+            cdf = scipy.stats.beta.cdf
+            pdf = scipy.stats.beta.pdf
+            eps = 0.01 * beta / (1.0 + np.sqrt(beta))
+            cdf_beta = (cdf(x, alpha, beta + eps) - cdf(x, alpha, beta - eps)) / (2 * eps)
+            cdf_x = pdf(x, alpha, beta)
+            expected_grad = -cdf_beta / cdf_x
+            rel_error = np.abs(actual_grad - expected_grad) / (expected_grad + 1e-30)
+            self.assertLess(np.max(rel_error), 0.005, '\n'.join([
+                'Bad gradient dx/dbeta for x ~ Beta({}, {})'.format(alpha, beta),
+                'x {}'.format(x),
+                'expected {}'.format(expected_grad),
+                'actual {}'.format(actual_grad),
+                'rel error {}'.format(rel_error),
+                'max error {}'.format(rel_error.max()),
+                'at x = {!r}'.format(x[rel_error.argmax()]),
+            ]))
+
+    def test_dirichlet_multivariate(self):
+        alpha_crit = 0.25 * (5.0 ** 0.5 - 1.0)
+        num_particles = 100000
+        for shift in [-0.1, -0.05, -0.01, 0.0, 0.01, 0.05, 0.10]:
+            alpha = alpha_crit + shift
+            alpha = Variable(torch.FloatTensor([alpha]), requires_grad=True)
+            alpha_vec = torch.cat([alpha, alpha, alpha.new([1])])
+            z = Dirichlet(alpha_vec.expand(num_particles, 3)).rsample()
+            mean_z3 = 1.0 / (2.0 * alpha + 1.0)
+            loss = torch.pow(z[:, 2] - mean_z3, 2.0).mean()
+            actual_grad = grad(loss, [alpha])[0].data
+            # Compute expected gradient by hand.
+            num = 1.0 - 2.0 * alpha - 4.0 * alpha**2
+            den = (1.0 + alpha)**2 * (1.0 + 2.0 * alpha)**3
+            expected_grad = (num / den).data
+            self.assertEqual(actual_grad, expected_grad, 0.002, '\n'.join([
+                "alpha = alpha_c + %.2g" % shift,
+                "expected_grad: %.5g" % expected_grad,
+                "actual_grad: %.5g" % actual_grad,
+                "error = %.2g" % torch.abs(expected_grad - actual_grad).max(),
+            ]))
+
+    def test_dirichlet_tangent_field(self):
```

Comment:
@martinjankowiak Could you please review this test?

Reply:
beautiful!